packages/evm/src/precompiles/12-execute.ts

```diff
@@ -0,0 +1,171 @@
+const SuffixDiff = ssz.container({
+  suffix: ssz.uint8,
+  currentValue: ssz.bytevector(32),
+  newValue: ssz.bytevector(32),
+})
+
+// For state diff entries
+const StateDiff = ssz.container({
+  stem: ssz.bytevector(31), // The stem as a hex string
+  suffixDiffs: ssz.list(256, SuffixDiff), // List of suffix diffs
+})
+
+// For proof entries
+const ProofEntry = ssz.container({
+  stem: ssz.bytevector(31), // 31-byte vector for the stem
+  proofData: ssz.list(32, ssz.bytelist(16384)), // List of byte arrays, each up to 16384 bytes
+})
+
+// Define the BinaryTreeExecutionWitness container
+const BinaryTreeExecutionWitness = ssz.container({
+  stateDiff: ssz.list(1024, StateDiff), // List of state diffs
+  parentStateRoot: ssz.bytevector(32), // Parent state root as hex
+  proof: ssz.list(256, ProofEntry), // List of proof entries with stems and proof data
+})
+
+const MAX_CALL_DATA_SIZE = 7500000 // Assuming a transaction with all zero bytes fills up an entire block worth of gas
+export const traceContainer: ssz.SSZCoder<any> = ssz.container({
+  txs: ssz.list(
+    // An ssz list of tx objects that match the `eth_call` tx object format
+    256,
+    ssz.container({
+      to: ssz.bytevector(20),
+      from: ssz.bytevector(20),
+      gasLimit: ssz.uint64,
+      gasPrice: ssz.uint64,
+      value: ssz.uint64,
+      data: ssz.bytelist(MAX_CALL_DATA_SIZE),
+    }),
+  ),
+  witness: BinaryTreeExecutionWitness,
+})
+
+export const stateWitnessJSONToSSZ = (
+  witness: Awaited<ReturnType<typeof generateBinaryExecutionWitness>>,
+) => {
+  return {
+    stateDiff: witness.stateDiff.map((diff) => ({
+      stem: hexToBytes(diff.stem),
+      suffixDiffs: diff.suffixDiffs.map((suffixDiff) => ({
+        suffix: suffixDiff.suffix,
+        currentValue:
+          suffixDiff.currentValue !== null
+            ? hexToBytes(suffixDiff.currentValue)
+            : new Uint8Array(32),
+        newValue:
+          suffixDiff.newValue !== null ? hexToBytes(suffixDiff.newValue) : new Uint8Array(32),
+      })),
+    })),
+    parentStateRoot: hexToBytes(witness.parentStateRoot),
+    proof: Object.entries(witness.proof).map(([stem, proof]) => ({
+      stem: hexToBytes(stem),
+      proofData: proof,
+    })),
+  }
+}
+
+export async function precompile12(opts: PrecompileInput): Promise<ExecResult> {
+  const pName = getPrecompileName('12')
+  const data = opts.data
+  const evm = opts._EVM as EVM
+  const gasUsed = opts.common.param('executeGasCost')
+  if (!gasLimitCheck(opts, gasUsed, pName)) {
+    return OOGResult(opts.gasLimit)
+  }
+  if (data.length !== 128) {
+    return EvmErrorResult(new EvmError(ERROR.INVALID_INPUT_LENGTH), opts.gasLimit)
+  }
+  const _preStateRoot = data.subarray(0, 32) // prestateroot for L2 state
+  const postStateRoot = data.subarray(32, 64) // post state root for L2 state
+  const traceBlob = evm['executionBlobs'].get(bytesToHex(data.subarray(64, 96))) // reference to state access and transactions
```

Comment:
so this is like versioned hash? only 1 traceblob is supported for now?

Reply:
Right, this is just a hack to reference the blob of data needed for re-executing the transactions until I figured out a clean way to access the data in an actual blob attached to the tx.